Extracting syn-SFS only

andrewkern opened the pull request:
Okay a bit of Github surgery had to happen but I believe I now how the commits from @izabelcavassim's #44 imported on to this clean branch and ready to run. 

This will need some review as we will need to double check everything works with different masks applied, but this should get us started. 

Individual commit messages follow 

-----------------
Update stairway.py

Fixed masks (finally)

Add non neutral SFS to plots

updated sfs plot

added some details

fixed filtering in msmc

Update config.yaml

Update config.yaml

fixed based on comments

sumarized mask script

created a function to mask

Update n_t.snake

andrewkern commented:
now that #67 has been merged, it will be interesting to rerun this analysis pipeline **with** masking.

I think it might be good if I added a commit on to this PR that has a mask file that is annotation specific, as currently masks are only basic on genetic map considerations I believe.

andrewkern commented on workflows/stairway.py:
```diff
@@ -75,6 +56,16 @@ def ts_to_stairway(self, ts_path, num_bootstraps=1, mask_file=None):
+                rows = (mask_table[0] == chrom)
+                mask_intervals = mask_table[rows].loc[:, 1:3].values
+                ts = ts.keep_intervals(mask_intervals)
             total_length += ts.sequence_length
```
i don't think we need to hold on to this for any reason?

andrewkern commented:
at this point masking is cleaned up and working for `stairwayplot`. this could be a natural place for a merge maybe?

andrewkern commented on workflows/smc.py:
```diff
@@ -36,10 +36,10 @@ def write_smcpp_file(path, mask_file, selection=None):
         subprocess.run(cmd, shell=True, check=True)
-        cmd = f"docker run --rm -v $PWD:/mnt terhorst/smcpp:latest vcf2smc --mask {inter_mask_path} {vz_file} "
+        cmd = f"smc++ vcf2smc --mask {inter_mask_path} {vz_file} "
         cmd = cmd + f"{smc_file} {chr_name} pop1:"
     else:
-        cmd = f"docker run --rm -v $PWD:/mnt terhorst/smcpp:latest vcf2smc {vz_file} {smc_file} {chr_name} pop1:"
+        cmd = f"smc++ vcf2smc {vz_file} {smc_file} {chr_name} pop1:"
```
these commits to `smc.py` came over from #44. we still need to decide once and for all if we will include `smc++` so i've decided to keep these changes here.

andrewkern commented:
i've gone ahead and added masking to `gone`. that should close #69. 

At this point masking based on genetic map considerations is complete for the `n_t.snake` workflow. We may want to expand this masking to cover all of the annotations used for selection.

andrewkern commented:
had a brief chat with @petrelharp about this today and decided that we will indeed want to mask annotations for at least some subset of the analysis, so i plan to implement this next on this branch.

andrewkern commented:
okay, masking is now extended on the N_t workflow to cover all annotations used during simulation as well as the low rec regions

petrelharp commented:
This PR does:
- mask out low recomb regions
- mask out exons, if provided
- optionally using only neutral mutation types

petrelharp commented on workflows/gone.py:
```diff
@@ -61,6 +62,14 @@ def ts2plink(ts,ped_file,map_file,gm_chr):
     # make .ped file
     genomat = ts.genotype_matrix() + 1 # it takes alleles (1,2)
+    # mask if appropriate
+    if type(mask_intervals) is not None:
```
```suggestion
    if mask_intervals is not None:
```

petrelharp commented on workflows/gone.py:
```diff
@@ -45,7 +46,7 @@ def copy(gone_code,outpath,seed,threads):
-def ts2plink(ts,ped_file,map_file,gm_chr):
+def ts2plink(ts, ped_file, map_file, gm_chr, chrID, mask_intervals=None):
     """
     converts ts to plink format
```
clarify whether mask is an inclusion or exclusion mask?

petrelharp commented on workflows/gone.py:
```diff
@@ -61,6 +62,14 @@ def ts2plink(ts,ped_file,map_file,gm_chr):
+    if type(mask_intervals) is not None:
+        positions = ts.tables.sites.position
+        for interval in mask_intervals:
+            masked_idx = np.where(np.logical_and(positions>=interval[0], positions<=interval[1]))[0]
```
```suggestion
            masked_idx = np.where(np.logical_and(positions>=interval[0], positions<interval[1]))[0]
```

petrelharp commented on workflows/gone.py:
```diff
@@ -61,6 +62,14 @@ def ts2plink(ts,ped_file,map_file,gm_chr):
     genomat = ts.genotype_matrix() + 1 # it takes alleles (1,2)
+    # mask if appropriate
+    if type(mask_intervals) is not None:
+        positions = ts.tables.sites.position
```
https://tskit.dev/tskit/docs/latest/python-api.html#tskit.TreeSequence.sites_position
```suggestion
        positions = ts.sites_position
```

petrelharp commented on workflows/msmc.py:
```diff
@@ -28,40 +28,34 @@ def getIntervalOverlap(intA, intB):
         fi = open(output, "w")
         prev = 0
-        if mask_file:
+        if type(mask_intervals) != None:
```
```suggestion
        if mask_intervals is not None:
```

petrelharp replied:
also note the stray print below?

petrelharp commented on workflows/msmc.py:
```diff
@@ -28,40 +28,34 @@ def getIntervalOverlap(intA, intB):
+        if type(mask_intervals) != None:
+            print(mask_intervals)
+            mask_intervals = pd.IntervalIndex.from_arrays(mask_intervals[:,0], mask_intervals[:,1])
             for var in ts.variants():
```
variants are in 1-to-1 correspondence with sites, and so always have unique positions: the `cur` and `prev` stuff here is unnecessary.

petrelharp replied:
sorry, I mean the `if cur > prev` condition

petrelharp replied:
ah except for the first position, if it's at site 0, so never mind - it's harmless

petrelharp commented on workflows/stairway.py:
```diff
@@ -55,53 +57,74 @@ def ts_to_stairway(self, ts_path, num_bootstraps=1, mask_file=None):
             ts = tskit.load(ts_p)
             total_length += ts.sequence_length
+            # masking?
+            if type(mask_intervals) != None:
```
```suggestion
            if mask_intervals is not None:
```

petrelharp commented on workflows/stairway.py:
```diff
@@ -55,53 +57,74 @@ def ts_to_stairway(self, ts_path, num_bootstraps=1, mask_file=None):
+            assert sum(site_class == None) == 0
+
+            # All SNP locs
+            snp_locs_all = ts.tables.sites.position.astype(int)
```
```suggestion
            snp_locs_all = ts.sites_position.astype(int)
```

petrelharp reviewed (commented):
Looks good except for these `type(x) != None`, which is always `False`.

xin-huang commented on workflows/masks.py:
```diff
@@ -0,0 +1,91 @@
+    :chrom_annotation: chromosome annotation name (e.g., "ensembl_havana_104_CDS")
+    :chromID: chromosome ID (e.g., "chr1")
+    """
+    species = stdpopsim.get_species("HomSap")
+    a = species.get_annotations("ensembl_havana_104_CDS")
+    mask = a.get_chromosome_annotations("chr21")
```
```suggestion
    species = stdpopsim.get_species(speciesID)
    a = species.get_annotations(chrom_annotation)
    mask = a.get_chromosome_annotations(chromID)
```

xin-huang commented on workflows/stairway.py:
```diff
@@ -36,15 +36,17 @@ class StairwayPlotRunner(object):
         self.workdir.mkdir(parents=True)
         stairway_path = pathlib.Path(stairway_dir)
-        self.classpath = "{}:{}".format(stairway_path, stairway_path / "swarmops.jar")
+        self.classpath = "{}:{}".format(
+            stairway_path, stairway_path / "swarmops.jar")
```
need to add `openjdk` into `environment.yml`

andrewkern replied:
added

xin-huang commented on workflows/stairway.py:
```diff
@@ -55,53 +57,74 @@ def ts_to_stairway(self, ts_path, num_bootstraps=1, mask_file=None):
-            plots.plot_sfs(SFSs, sfs_path)
-            # Bootstrap allele counts
+
+            # Mapping mutation type IDs to class of mutation (e.g., neutral, non-neutral)
+            class_muts = {}
+            for dfe in ts.metadata["stdpopsim"]["DFEs"]:
+                for mt in dfe["mutation_types"]:
+                    mid = mt["slim_mutation_type_id"]
+                    if not mid in class_muts:
+                        class_muts[mid] = "neutral" if mt["is_neutral"] else "non_neutral"
+
+            site_class = np.empty(ts.num_sites, dtype=object)
+
+            # Finding neutral positions
+            neu_positions = []
+            non_neu_positions = []
+            for j, s in enumerate(ts.sites()):
+                mut_hits = []
+                for m in s.mutations:
+                    for md in m.metadata["mutation_list"]:
+                        mut_hits.append(md["mutation_type"])
+                        if set(class_muts[md["mutation_type"]]) == set("neutral"):
+                            neu_positions.append(m.site)
+                        if set(class_muts[md["mutation_type"]]) == set("non_neutral"):
+                            non_neu_positions.append(m.site)
+                site_class[j] = class_muts[mut_hits[0]] if len(
+                    mut_hits) == 1 else "double_hit"
+            assert sum(site_class == None) == 0
```
Can we reuse the codes `_generate_fs_from_ts` in `ts2fs.py`?

andrewkern replied:
there's probably a way to do this but it would take a bit of coding as currently `_generate_fs_from_ts` returns an array of allele frequency spectra for each class of mutation. this could be a good thing to wrap in a function however, and would be a simple enough PR

andrewkern commented on workflows/config/snakemake/config.yaml:
```diff
@@ -4,41 +4,46 @@
+    ]
+
+"genetic_map": "HapMapII_GRCh38"
 "chrm_list": "chr21"
 "dfe_list": ["Gamma_H17"]
 "output_dir": "results"
-
+"mask_file": "workflows/masks/HapmapII_GRCh37.mask.bed"
```
yes we will need to rerun the masking scripts to get the new coordinates, or alternatively we might just lift these over to hg38

andrewkern commented:
okay that's for the code review @petrelharp and @xin-huang. i've now incorporated all these edits and squashed. i'm going to go ahead and now merge this and we can keep building.
